## 1. Import Required Libraries

In [1]:
import pandas as pd
import numpy as np

## 2. Load the Dataset

In [ ]:
df = pd.read_csv("../00-dataSource/dataSourceV2/raw_data_v2.csv")

,User_ID,Age,Gender,Daily_Screen_Time(hrs),Sleep_Quality(1-10),Stress_Level(1-10),Days_Without_Social_Media,Exercise_Frequency(week),Social_Media_Platform,Happiness_Index(1-10)
0,U001,44,Male,3.1,7.0,6.0,2.0,5.0,Facebook,10.0
1,U002,30,Other,5.1,7.0,8.0,5.0,3.0,LinkedIn,10.0
2,U003,23,Other,7.4,6.0,7.0,1.0,3.0,YouTube,6.0
3,U004,36,Female,5.7,7.0,8.0,1.0,1.0,TikTok,8.0
4,U005,34,Female,7.0,4.0,7.0,5.0,1.0,X (Twitter),8.0


## 3. Initial Data Exploration

In [5]:
# Show first 5 rows: quick preview of dataset structure and values
print("First 5 rows:")
display(df.head())

# Show dataset size: number of rows and columns
print("Shape of dataset:")
print(df.shape)

# Show column names: list of all features
print("\nColumn names:")
print(df.columns)

# Show index: how rows are labeled (usually default range)
print("\nIndex:")
print(df.index)

# Show numerical summary: mean, min, max, etc.
print("\nStatistical summary (Numerical):")
display(df.describe())

# Show data types and non-null counts: detect missing values & wrong types
print("\nDataset information:")
df.info()

# Show missing values per column: identify incomplete data
print("\nMissing values:")
display(df.isnull().sum())

# Show number of duplicate rows: detect repeated records
print("\nDuplicate rows:")
print(df.duplicated().sum())

First 5 rows:


,User_ID,Age,Gender,Daily_Screen_Time(hrs),Sleep_Quality(1-10),Stress_Level(1-10),Days_Without_Social_Media,Exercise_Frequency(week),Social_Media_Platform,Happiness_Index(1-10)
0,U001,44,Male,3.1,7.0,6.0,2.0,5.0,Facebook,10.0
1,U002,30,Other,5.1,7.0,8.0,5.0,3.0,LinkedIn,10.0
2,U003,23,Other,7.4,6.0,7.0,1.0,3.0,YouTube,6.0
3,U004,36,Female,5.7,7.0,8.0,1.0,1.0,TikTok,8.0
4,U005,34,Female,7.0,4.0,7.0,5.0,1.0,X (Twitter),8.0


Shape of dataset:
(500, 10)

Column names:
Index(['User_ID', 'Age', 'Gender', 'Daily_Screen_Time(hrs)',
       'Sleep_Quality(1-10)', 'Stress_Level(1-10)',
       'Days_Without_Social_Media', 'Exercise_Frequency(week)',
       'Social_Media_Platform', 'Happiness_Index(1-10)'],
      dtype='str')

Index:
RangeIndex(start=0, stop=500, step=1)

Statistical summary (Numerical):


,Age,Daily_Screen_Time(hrs),Sleep_Quality(1-10),Stress_Level(1-10),Days_Without_Social_Media,Exercise_Frequency(week),Happiness_Index(1-10)
count,500.000000,500.000000,500.000000,500.000000,500.000000,500.000000,500.000000
mean,32.988000,5.530000,6.304000,6.618000,3.134000,2.448000,8.376000
std,9.960637,1.734877,1.529792,1.542996,1.858751,1.428067,1.524228
min,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,4.000000
25%,24.000000,4.300000,5.000000,6.000000,2.000000,1.000000,7.000000
50%,34.000000,5.600000,6.000000,7.000000,3.000000,2.000000,9.000000
75%,41.000000,6.700000,7.000000,8.000000,5.000000,3.000000,10.000000
max,49.000000,10.800000,10.000000,10.000000,9.000000,7.000000,10.000000



Dataset information:
<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   User_ID                    500 non-null    str    
 1   Age                        500 non-null    int64  
 2   Gender                     500 non-null    str    
 3   Daily_Screen_Time(hrs)     500 non-null    float64
 4   Sleep_Quality(1-10)        500 non-null    float64
 5   Stress_Level(1-10)         500 non-null    float64
 6   Days_Without_Social_Media  500 non-null    float64
 7   Exercise_Frequency(week)   500 non-null    float64
 8   Social_Media_Platform      500 non-null    str    
 9   Happiness_Index(1-10)      500 non-null    float64
dtypes: float64(6), int64(1), str(3)
memory usage: 39.2 KB

Missing values:


User_ID                      0
Age                          0
Gender                       0
Daily_Screen_Time(hrs)       0
Sleep_Quality(1-10)          0
Stress_Level(1-10)           0
Days_Without_Social_Media    0
Exercise_Frequency(week)     0
Social_Media_Platform        0
Happiness_Index(1-10)        0
dtype: int64


Duplicate rows:
0


## Summary of the Dataset

From the initial exploration, the dataset contains **500 rows** and **10 columns**, meaning there are 500 observations and 10 features describing users’ digital lifestyle, behavior, and well-being.

## Dataset Structure

| Category              | Details |
|----------------------|--------|
| Number of Rows       | 500 |
| Number of Columns    | 10 |
| Numerical Features   | 7 |
| Categorical Features | 3 |
| Missing Values       | None |
| Duplicate Rows       | 0 |

The dataset includes both **numerical** and **categorical** variables. Numerical features represent measurable values such as screen time, sleep quality, and stress level, while categorical features represent grouped information such as gender and social media platform.

## Feature Types

### **Numerical Features**
- Age  
- Daily_Screen_Time(hrs)  
- Sleep_Quality(1-10)  
- Stress_Level(1-10)  
- Days_Without_Social_Media  
- Exercise_Frequency(week)  
- Happiness_Index(1-10)  

### **Categorical Features**
- User_ID  
- Gender  
- Social_Media_Platform  

## Key Observations

- The dataset is **complete**, with no missing values in any column.  
- There are **no duplicate rows**, meaning each record is unique.  
- The data types are properly structured into:
  - numerical (`int64`, `float64`)
  - categorical (`str`)  
- The dataset focuses on:
  - digital behavior (screen time, social media usage)  
  - health and lifestyle (sleep quality, exercise)  
  - mental well-being (stress level, happiness index)  


## Statistical Insights

| Feature                          | Mean | Min | Max |
|----------------------------------|------|-----|-----|
| Age                              | 32.99 | 16 | 49 |
| Daily Screen Time (hrs)          | 5.53 | 1.0 | 10.8 |
| Sleep Quality (1–10)             | 6.30 | 2.0 | 10.0 |
| Stress Level (1–10)              | 6.62 | 2.0 | 10.0 |
| Days Without Social Media        | 3.13 | 0.0 | 9.0 |
| Exercise Frequency (per week)    | 2.45 | 0.0 | 7.0 |
| Happiness Index (1–10)           | 8.38 | 4.0 | 10.0 |


### Insights:
- Users spend an average of **5.5 hours per day on screens**  
- Sleep quality is moderate, averaging around **6.3/10**  
- Stress levels are relatively **high (average ≈ 6.6/10)**  
- Happiness levels are generally **high (average ≈ 8.4/10)**  
- Exercise frequency is relatively **low (≈ 2–3 days/week)**  

## 4. Data Cleaning

### 4.1 Handling Missing Values

All columns in the dataset contain **0 missing values**, which means the dataset is complete.

No imputation or removal is required for missing data.

In [6]:
# Check missing values in each column
print("Missing values:")
display(df.isnull().sum())

Missing values:


User_ID                      0
Age                          0
Gender                       0
Daily_Screen_Time(hrs)       0
Sleep_Quality(1-10)          0
Stress_Level(1-10)           0
Days_Without_Social_Media    0
Exercise_Frequency(week)     0
Social_Media_Platform        0
Happiness_Index(1-10)        0
dtype: int64

### 4.2 Handling Duplicate Records

There are **no duplicate rows** in the dataset.

This ensures that each record is unique and no bias is introduced from repeated data.

In [7]:
# Check duplicate rows
print("Duplicate rows:")
print(df.duplicated().sum())

Duplicate rows:
0


### 4.3 Data Type Verification

The dataset contains:
- numerical data (int64, float64)
- categorical data (string)

The data types appear correct and suitable for analysis. No conversion is required at this stage.

In [8]:
# Check data types
print("Data types:")
df.dtypes

Data types:


User_ID                          str
Age                            int64
Gender                           str
Daily_Screen_Time(hrs)       float64
Sleep_Quality(1-10)          float64
Stress_Level(1-10)           float64
Days_Without_Social_Media    float64
Exercise_Frequency(week)     float64
Social_Media_Platform            str
Happiness_Index(1-10)        float64
dtype: object

### 4.4 Check Logical Consistency

we divide it into two
- Categorical feature(Text, String)
- Numerical feature 

In [10]:
# categorical feature

# check unique value 
for col in ['Gender','Social_Media_Platform']:
    print(col, df[col].unique())

# check inconsistent format 
for col in ['Gender','Social_Media_Platform']:
    df[col] = df[col].str.strip().str.lower()

Gender <StringArray>
['Male', 'Other', 'Female']
Length: 3, dtype: str
Social_Media_Platform <StringArray>
['Facebook', 'LinkedIn', 'YouTube', 'TikTok', 'X (Twitter)', 'Instagram']
Length: 6, dtype: str


In [11]:
# Screen time (0–24 hours)
(df['Daily_Screen_Time(hrs)'] < 0).sum()
(df['Daily_Screen_Time(hrs)'] > 24).sum()

# Sleep quality (1–10 scale)
(df['Sleep_Quality(1-10)'] < 1).sum()
(df['Sleep_Quality(1-10)'] > 10).sum()

# Stress level (1–10 scale)
(df['Stress_Level(1-10)'] < 1).sum()
(df['Stress_Level(1-10)'] > 10).sum()

# Days without social media (should be ≥ 0)
(df['Days_Without_Social_Media'] < 0).sum()

# Exercise frequency (0–7 days/week)
(df['Exercise_Frequency(week)'] < 0).sum()
(df['Exercise_Frequency(week)'] > 7).sum()

# Happiness index (1–10 scale)
(df['Happiness_Index(1-10)'] < 1).sum()
(df['Happiness_Index(1-10)'] > 10).sum()

# Age (reasonable range check, e.g., 10–100)
(df['Age'] < 10).sum()
(df['Age'] > 100).sum()

np.int64(0)

### 4.5 remove some features that are not important

Several features were removed to improve model performance and avoid data leakage.

- The id column was removed as it does not provide useful information.
The remaining features include relevant behavioral and contextual variables for prediction.

In [12]:
df = df.drop(columns=[
    'User_ID',
])

### 4.6 Outlier Detection

In this step, we detect outliers in numerical features.

Outliers are extreme values that differ significantly from other observations.

They may represent:
- real but rare behavior
- data entry errors

Outliers are detected using the IQR (Interquartile Range) method.

In [13]:
# Select numerical columns
num_cols = df.select_dtypes(include=['int64', 'float64']).columns

# Function to count outliers
def count_outliers(df, col):
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    return ((df[col] < lower) | (df[col] > upper)).sum()

# Check outliers for each column
for col in num_cols:
    print(f"{col}: {count_outliers(df, col)} outliers")

Age: 0 outliers
Daily_Screen_Time(hrs): 2 outliers
Sleep_Quality(1-10): 0 outliers
Stress_Level(1-10): 1 outliers
Days_Without_Social_Media: 0 outliers
Exercise_Frequency(week): 2 outliers
Happiness_Index(1-10): 0 outliers


### Outlier Detection Summary

Outliers were identified using the **IQR method** across numerical features.

| Feature                          | Outliers |
|----------------------------------|----------|
| Age                              | 0 |
| Daily_Screen_Time(hrs)           | 2 |
| Sleep_Quality(1-10)              | 0 |
| Stress_Level(1-10)               | 1 |
| Days_Without_Social_Media        | 0 |
| Exercise_Frequency(week)         | 2 |
| Happiness_Index(1-10)            | 0 |

Most features show **little to no outliers**, indicating that the dataset is generally clean and well-distributed.

### Interpretation & Decision

Outliers are mainly found in behavioral features such as:
- daily screen time  
- exercise frequency  
- stress level  

These values are **not removed** because:
- they represent **realistic human behavior** (e.g., high screen usage or irregular exercise habits)  
- they may contain **important patterns for predicting stress and happiness**  
- no **invalid or impossible values** were detected during validation  

Removing these outliers may lead to **loss of useful information** and could negatively impact model performance.

## Copy clean data to csv

In [14]:
# Save cleaned dataset
df.to_csv("../00-dataSource/dataSourceV2/clean_data_v1.csv", index=False)

print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!
